#40 Implement Quick Outline: better labels

ghillairet opened the pull request:
This PR extends the current quick outline with better labels(#40) and introduces a new outline that overrides the one provided by YEdit with content identical to the quick outline.
The (quick) outline uses the unified model to represent the content of swagger documents.

ghillairet committed:
[#40] Implement Quick Outline

Changed the behaviour of quick outline selection, now tries to select the first line corresponding to the the selected element
instead of the whole element.
Added tests for labels of quick outline elements.

tedepstein committed:
Update README.md

ghillairet committed:
This commit introduces a novel approach for parsing of yaml documents. It uses Jackson streaming API to build a model representation of the content of the document. This model more or less matches the JSON structure, but includes specific information needed by SwagEdit such as location details for each nodes.
The introduction of this model is done to unify the representation of the document content between the different services provided by the editor, such as content assist, outline, quick outline, etc...

ghillairet committed:
Implement outline that overrides YEdit outline by providing content specific to SwagEdit.

ghillairet committed:
[#40] Implement Quick Outline

QuickOutline and Outline now uses a unified model to get access to the content of the document. The model provides a representation of the
document in form of nodes (object - array - value). Each node can be retrieve via a pointer (JSON pointer) or it's position inside the document.
Each node stores it's start and end position. Each node has access to it's children and parent, making it useful to be used in a tree content provider,
and can provide a label useful for label providers.

This commit also introduces better labels for the (quick)outline. Those labels are now derived from the type of elements.

ghillairet committed:
Merge pull request #176 from RepreZen/task/40-model-nodes

Task/40 model nodes

tfesenko committed:
Merge branch 'master' into task/40

tfesenko commented on com.reprezen.swagedit.tests/src/com/reprezen/swagedit/editor/outline/AbstractNodeTest.xtend:
```diff
@@ -0,0 +1,385 @@
+		assertEquals(0, doc.getLineOfOffset(position.offset))
+	}
+	
+//	@Test
```
The `@Test` annotation is comment. Is it because the implementation of the **test** is incomplete or because of  the code? Any reason for not using `@Ignore`?

tfesenko commented on com.reprezen.swagedit.tests/src/com/reprezen/swagedit/editor/outline/AbstractNodeTest.xtend:
```diff
@@ -0,0 +1,385 @@
+		assertEquals(0, position.length)
+
+		// should be first line
+		assertEquals(0, doc.getLineOfOffset(position.offset))
```
Thanks!

tfesenko commented on com.reprezen.swagedit.tests/src/com/reprezen/swagedit/editor/outline/OutlineStyledLabelProviderTest.xtend:
```diff
@@ -44,18 +42,62 @@ class OutlineStyledLabelProviderTest {
+		'''
+		
+		val els = Model.parseYaml(text).root
+		assertEquals("objects", els.get(0).text)
```
Nice

tfesenko commented on com.reprezen.swagedit.tests/src/com/reprezen/swagedit/editor/outline/OutlineStyledLabelProviderTest.xtend:
```diff
@@ -44,18 +42,62 @@ class OutlineStyledLabelProviderTest {
+		val els = Model.parseYaml(text).root
+		assertEquals("objects", els.get(0).text)
+		
+		// TODO
```
Let's create an issue for the TODO items, otherwise we can easily forget about them

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/outline/OutlineStyledLabelProvider.java:
```diff
@@ -56,35 +59,65 @@ public Styler getTextStyler() {
     }
 
-    protected StyledString getSyledString(OutlineElement element) {
+    protected StyledString getSyledString(AbstractNode element) {
```
Typo: a missing `t` in `getSyledString`

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/outline/QuickOutline.java:
```diff
@@ -121,6 +122,7 @@ protected TreeViewer createTreeViewer(Composite parent) {
         treeViewer.addFilter(new NamePatternFilter());
         // Using ALL_LEVELS will cause editor to hang on large specs
         treeViewer.setAutoExpandLevel(2);
```
Good. Let's extract the `AutoExpandLevel` into a constant

tfesenko committed:
#40 - select the entire first line of element in the editor when an
quick outline item is selected

Previously navigation did not work because `length` was 0

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/model/AbstractNode.java:
```diff
@@ -0,0 +1,150 @@
+                return new Position(0);
+            }
+
+            return new Position(Math.max(0, offset), 0);
```
Because the length (the second parameter) is always null navigation did not work and Quick Outline was basically useless. I provided a simple fix in commit d90ea00390076f38d982d68df1adb37623698b2e which selects the entire first line of an element:
<img width="680" alt="screen shot 2016-08-11 at 4 21 53 pm" src="https://cloud.githubusercontent.com/assets/644582/17603566/c3f456b0-5fdf-11e6-96c2-9cc02497f238.png">
@ghillairet please review update my fix and update if needed.

tedepstein replied:
Very glad to see that it's the property name line selected. That's very clean from UX perspective, takes the user exactly where he wanted to go.

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/model/NodeDeserializer.java:
```diff
@@ -0,0 +1,84 @@
+            node.setEndLocation(p.getCurrentLocation());
+            return node;
+
+        } else if (p.getCurrentToken() == JsonToken.START_ARRAY) {
```
What about extracting methods to handle `START_OBJECT`,  `START_ARRAY` and the default case instead of keeping them all inside a 70-line long method?

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/model/NodeDeserializer.java:
```diff
@@ -0,0 +1,84 @@
+            throws IOException, JsonProcessingException {
+
+        final Model model = (Model) context.getAttribute("model");
+        final AbstractNode parent = (AbstractNode) context.getAttribute("parent");
```
The magic strings `parent` and `pointer` are used at least three times. What about converting them into reusable constants?

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/model/AbstractNode.java:
```diff
@@ -0,0 +1,150 @@
+    private JsonType type;
+    private JsonNode schema;
+
+    public final Set<SchemaDefinition> definitions;
```
`definitions` is public, but there is a corresponding `getDefinitions()`. Was the field made `public` by mistake?

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/model/AbstractNode.java:
```diff
@@ -0,0 +1,150 @@
+        this.pointer = ptr;
+        this.location = location;
+        this.definitions = provider.getDefinitions(pointer(ptr));
+        this.schema = definitions.isEmpty() ? null : Iterables.getFirst(definitions, null).definition;
```
`getFirst(definitions, null).definition` looks like a potential NPE

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/model/AbstractNode.java:
```diff
@@ -0,0 +1,150 @@
+        return Collections.emptyList();
+    }
+
+    public abstract String getText();
```
I suggest to move this abstract method next to other abstract methods defined above to make it easier to spot all of them at once

tfesenko replied:
I just noticed that `ArrayNode` and `ObjectNode` have the same implementation of `getText()`:

``` java
return getProperty() == null ? "" : getProperty();
```

Only `ValueNode` has different implementation.
@ghillairet you can use it as a default `getText()` and override it for `ValueNode`

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/model/AbstractNode.java:
```diff
@@ -0,0 +1,150 @@
+        return property;
+    }
+
+    public void setProperty(String name) {
```
is it propertyName or propertyValue? Based on the parameter, it's more like `propertyName`, is it correct? Can we rename the method or parameter name then?

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/model/ArrayNode.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Override
+    public String getText() {
+        return getProperty() == null ? "" : getProperty();
```
What about 

``` java
Optional.<String>of(getProperty()).or("")
```

?

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/outline/SwaggerContentOutlinePage.java:
```diff
@@ -0,0 +1,66 @@
+import com.reprezen.swagedit.editor.SwaggerDocument;
+import com.reprezen.swagedit.model.Model;
+
+public class SwaggerContentOutlinePage extends ContentOutlinePage {
```
Nice

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/outline/SwaggerContentOutlinePage.java:
```diff
@@ -0,0 +1,66 @@
+        viewer.setContentProvider(new OutlineContentProvider());
+        viewer.setLabelProvider(new OutlineStyledLabelProvider());
+        viewer.addSelectionChangedListener(this);
+        viewer.setAutoExpandLevel(2);
```
I found a bug which may be related to `setAutoExpandLevel(2)`
I noticed that if I expand an element in the Outline View, then edit the element in the editor, the corresponding node in the outline which was previously expanded gets collapsed:
![outline_collapse](https://cloud.githubusercontent.com/assets/644582/17605090/b0413014-5fe6-11e6-95fe-df6f0b7619b4.gif)

It's a bug.

ghillairet replied:
PR for this bug https://github.com/RepreZen/SwagEdit/pull/185
